### PR TITLE
at_cmd and at_host checks for empty commands

### PR DESCRIPTION
**Pull request by bjda**
Adds checks for empty commands to react as expected by AT clients. Calls to at_cmd with such commands will fail properly, and the AT client will drop them silently.

### CODE REVIEW

File: lib/at_cmd/at_cmd.c [hunk 1]
```diff
@@ -76,6 +76,24 @@ static int open_socket(void)
...
+/*
+ * Do any validation of an AT command not performed by the lower layers or
+ * by the modem.
+ */
+static int check_cmd(const char *cmd) {
+	if (cmd == NULL) {
+		return -EINVAL;
+	}
+
+	// Check for presence of at least one printable non-whitespace character
```
Comment (rlubos) [inline lib/at_cmd/at_cmd.c:88]: Please use C-style comments.

File: lib/at_cmd/at_cmd.c [hunk 2]
```diff
@@ -76,6 +76,24 @@ static int open_socket(void)
 	return 0;
 }
 
+/*
+ * Do any validation of an AT command not performed by the lower layers or
+ * by the modem.
+ */
+static int check_cmd(const char *cmd) {
```
Comment (rlubos) [inline lib/at_cmd/at_cmd.c:83]: Opening bracket shall be put in a separate line (only for function definitions).

File: lib/at_cmd/at_cmd.c [hunk 3]
```diff
@@ -76,6 +76,24 @@ static int open_socket(void)
...
+	if (cmd == NULL) {
+		return -EINVAL;
+	}
+
+	// Check for presence of at least one printable non-whitespace character
+	for (char *c = cmd; c++; *c != '\0') {
```
Comment (rlubos) [inline lib/at_cmd/at_cmd.c:89]: Shouldn't this be `for (char *c = cmd; *c != '\0'; c++) {`? `*c != '\0'` have no use here currently.

File: lib/at_host/at_host.c [hunk 4]
```diff
@@ -182,6 +182,15 @@ static void uart_rx_handler(uint8_t character)
 	inside_quotes = false;
 	at_cmd_len = 0;
 
+	// Check for presence of at least one printable non-whitespace character
```
Comment (rlubos) [inline lib/at_host/at_host.c:185]: C-style comments please, here and below.

File: lib/at_host/at_host.c [hunk 5]
```diff
@@ -182,6 +182,15 @@ static void uart_rx_handler(uint8_t character)
...
+	for (char *c = at_buf; c++;) {
+		if (*c > ' '){
```
Comment (rlubos) [inline lib/at_host/at_host.c:187]: Missing space before `{`.

**Comment by richardmccrae-nordicsemi** [comment]
I have run my tests on this PR with the following results:
The empty commands "" and " " returns -EINVAL as expected, no freezing.  Thumbs up
Commands such as "/n/r", "/r/n", "/r/n/r/r/r/n/n/n/r/n", etc return 0.  Should these also return -EINVAL or is 0 correct here?

File: lib/at_cmd/at_cmd.c (hunk 3 above)
Comment (bjda) [reply lib/at_cmd/at_cmd.c]: Fixed. This would have been breaking, thanks for pointing it out.

**Comment by bjda** [comment]
> 
> 
> I have run my tests on this PR with the following results:
> The empty commands "" and " " returns -EINVAL as expected, no freezing. Thumbs up
> Commands such as "/n/r", "/r/n", "/r/n/r/r/r/n/n/n/r/n", etc return 0. Should these also return -EINVAL or is 0 correct here?

@richardmccrae-nordicsemi That is odd, are you sure the control characters are sent, and not literal slashes r's and n's? The code should respond similarly to all control characters and space..

**Comment by bjda** [comment]
@rlubos The formatting issues and typo are now fixed.

**Review by rlubos — APPROVED** [review]

**Review by richardmccrae-nordicsemi — APPROVED** [review]

File: lib/at_cmd/at_cmd.c [hunk 6]
```diff
@@ -76,6 +76,25 @@ static int open_socket(void)
...
+ * by the modem.
+ */
+static int check_cmd(const char *cmd)
+{
+	if (cmd == NULL) {
+		return -EINVAL;
+	}
+
+	/* Check for the presence one printable non-whitespace character */
+	for (const char *c = cmd; *c != '\0'; c++) {
```
Comment (evenl) [inline lib/at_cmd/at_cmd.c:90]: This looks a bit unsafe, relying only on 0-termination. Do you really need to check the whole string, isn't enough to only check the first character? And fail the string if begin with a whitespace character. In the at command parser library we have some utility commands to detect different parts of the AT strings. Maybe these functions can be re-used in this context.

This expression will also accept all of the extended ascii characters, which I am not sure are valid AT string characters.

**Pull request merged**